File: backend/src/main/java/com/unbora/api/domain/notification/NotificationsService.java
package com.unbora.api.domain.notification;

import com.unbora.api.common.LocationsConstants;
import com.unbora.api.common.exception.ApiException;
import com.unbora.api.domain.notification.dto.CreateNotificationDto;
import com.unbora.api.domain.notification.dto.NotificationRecordDto;
import com.unbora.api.domain.notification.dto.UpdateNotificationDto;
import org.springframework.http.HttpStatus;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.Instant;
import java.util.List;
import java.util.Map;
import java.util.UUID;

@Service
public class NotificationsService {

    private final NotificationRepository notificationRepository;

    public NotificationsService(NotificationRepository notificationRepository) {
        this.notificationRepository = notificationRepository;
    }

    private NotificationRecordDto toRecord(Notification n) {
        return new NotificationRecordDto(
                n.getId(),
                n.getTitle(),
                n.getBody(),
                n.getCity() != null ? n.getCity() : LocationsConstants.DEFAULT_CITY,
                n.getRegion() != null ? n.getRegion() : LocationsConstants.DEFAULT_REGION,
                n.getActive(),
                n.getCreatedAt() != null ? n.getCreatedAt().toString() : "",
                n.getUpdatedAt() != null ? n.getUpdatedAt().toString() : ""
        );
    }

    public List<NotificationRecordDto> list(boolean activeOnly, String city, String region) {
        return notificationRepository.findAllByOrderByCreatedAtDesc().stream()
                .filter(n -> !activeOnly || Boolean.TRUE.equals(n.getActive()))
                .filter(n -> LocationsConstants.matchesLocation(n.getCity(), n.getRegion(), city, region))
                .map(this::toRecord)
                .toList();
    }

    @Transactional
    public NotificationRecordDto create(CreateNotificationDto dto) {
        Notification notification = new Notification();
        notification.setId(UUID.randomUUID().toString());
        notification.setTitle(dto.title().trim());
        notification.setBody(dto.body().trim());
        notification.setCity(dto.city().trim());
        notification.setRegion(dto.region().trim());
        notification.setActive(dto.active() != null ? dto.active() : true);
        notification.setCreatedAt(Instant.now());
        notification.setUpdatedAt(Instant.now());

        return toRecord(notificationRepository.save(notification));
    }

    @Transactional
    public NotificationRecordDto update(String id, UpdateNotificationDto dto) {
        Notification notification = notificationRepository.findById(id)
                .orElseThrow(() -> new ApiException(HttpStatus.NOT_FOUND, "Notificação não encontrada."));

        if (dto.title() != null && !dto.title().isBlank()) notification.setTitle(dto.title().trim());
        if (dto.body() != null && !dto.body().isBlank()) notification.setBody(dto.body().trim());
        if (dto.city() != null && !dto.city().isBlank()) notification.setCity(dto.city().trim());
        if (dto.region() != null && !dto.region().isBlank()) notification.setRegion(dto.region().trim());
        if (dto.active() != null) notification.setActive(dto.active());
        notification.setUpdatedAt(Instant.now());

        return toRecord(notificationRepository.save(notification));
    }

    @Transactional
    public Map<String, Object> remove(String id) {
        Notification notification = notificationRepository.findById(id)
                .orElseThrow(() -> new ApiException(HttpStatus.NOT_FOUND, "Notificação não encontrada."));

        notificationRepository.delete(notification);
        return Map.of("deleted", true, "id", id);
    }
}

File: backend-nest-backup/src/notifications/notifications.service.ts
import { Injectable, NotFoundException } from '@nestjs/common';
import { Notification, Prisma } from '@prisma/client';
import { randomUUID } from 'crypto';

import {
  DEFAULT_CITY,
  DEFAULT_REGION,
  matchesLocation,
  normalizeCity,
  normalizeRegion,
} from '../common/locations';
import { PrismaService } from '../prisma/prisma.service';

import {
  CreateNotificationDto,
  NotificationRecord,
  UpdateNotificationDto,
} from './notifications.types';

@Injectable()
export class NotificationsService {
  constructor(private readonly prisma: PrismaService) {}

  private toRecord(item: Notification): NotificationRecord {
    return {
      id: item.id,
      title: item.title,
      body: item.body,
      city: item.city ?? DEFAULT_CITY,
      region: item.region ?? DEFAULT_REGION,
      active: item.active,
      createdAt: item.createdAt.toISOString(),
      updatedAt: item.updatedAt.toISOString(),
    };
  }

  async list(options: {
    activeOnly?: boolean;
    city?: string;
    region?: string;
  } = {}): Promise<NotificationRecord[]> {
    const city = normalizeCity(options.city);
    const region = normalizeRegion(options.region);

    const items = await this.prisma.notification.findMany({
      where: {
        ...(options.activeOnly ? { active: true } : {}),
        ...(city ? { city: { equals: city, mode: 'insensitive' } } : {}),
        ...(region ? { region: { equals: region, mode: 'insensitive' } } : {}),
      },
      orderBy: { createdAt: 'desc' },
    });

    return items
      .map((item) => this.toRecord(item))
      .filter((item) => !city && !region ? true : matchesLocation(item, city, region));
  }

  async create(dto: CreateNotificationDto): Promise<NotificationRecord> {
    const notification = await this.prisma.notification.create({
      data: {
        id: randomUUID(),
        title: dto.title.trim(),
        body: dto.body.trim(),
        city: dto.city.trim(),
        region: dto.region.trim(),
        active: dto.active ?? true,
      },
    });

    return this.toRecord(notification);
  }

  async update(id: string, dto: UpdateNotificationDto): Promise<NotificationRecord> {
    try {
      const data: Prisma.NotificationUpdateInput = {};

      if (dto.title !== undefined) data.title = dto.title.trim();
      if (dto.body !== undefined) data.body = dto.body.trim();
      if (dto.city !== undefined) data.city = dto.city.trim();
      if (dto.region !== undefined) data.region = dto.region.trim();
      if (dto.active !== undefined) data.active = dto.active;

      const updated = await this.prisma.notification.update({ where: { id }, data });
      return this.toRecord(updated);
    } catch {
      throw new NotFoundException('Notificação não encontrada.');
    }
  }

  async remove(id: string): Promise<{ deleted: true; id: string }> {
    try {
      await this.prisma.notification.delete({ where: { id } });
      return { deleted: true, id };
    } catch {
      throw new NotFoundException('Notificação não encontrada.');
    }
  }
}
